edx/edx-enterprise
# ENT-76: Make sure IdP and EC have OneToOne Relation

@mattdrayer , @e-kolpakov , @bradenmacdonald , @asadiqbal08 -- FYI

Kindly review this PR, it adds a `EnterpriseCustomerIdentityProvider` model and corresponding admin interface. 

JIRA: [ENT-76](https://openedx.atlassian.net/browse/ENT-76)

__Description:__
Currently, identity provider is a Slug field which can be duplicated across enterprise customers, which presents a problem for LMS apps that need to access enterprise customer for a given identity provider.
We need to make sure that enterprise customer and identity provider has a one to one relation.

__Testing Instructions:__
1. Go to SAML Provider Config admin page at "admin/third_party_auth/samlproviderconfig/" and add a couple of new identity providers. (make sure that the site selected for identity provider is same as the one associated with your current domain)
1. Go to enterprise customer admin portal at "/admin/enterprise/enterprisecustomer/"
    - Add a new enterprise customer without selecting any identity provider from the drop down. An enterprise customer should be added in the system, make sure that new enterprise customer does not have any associated identity provider.
    - Try adding another enterprise customer, this time select an identity provider from the dropdown  An enterprise customer should be added in the system, make sure that new enterprise customer have correct associated identity provider.
    -  Try adding another enterprise customer, select a site that is not the same as identity provider's site (this is the site that you selected in step 1). There should be an error message explaining site mismatch between enterprise customer and identity provider.
    - Try adding another Enterprise Customer with an identity provider that is already associated with another enterprise customer. There should be an error message saying that the selected identity provider is already associated with another enterprise customer.
    - Try removing an associated identity provider from an existing enterprise customer, you can do that by opening enterprise customer details view and selecting `delete` checkbox on the right corner of "ENTERPRISE CUSTOMER IDENTITY PROVIDERS" inline form and then hitting save.
